Support for generic templates

srvsh opened the pull request:
1. Moved template information out of `p5-app-duckpan` and into respective IA packages
2. The templates are now defined in YAML files. See https://github.com/srvsh/zeroclickinfo-goodies/blob/3a4f248f61cdb343df37f0d84861129d91c9de7f/template/templates.yml for an example. The file also has some documentation about the structure of the YAML file.
3. An instant answer can have multiple templates that can be chosen by the user during creation.
4. A template has some required files and some optional files. Optional files are in groups and users must specify which of these groups of files they'd like to generate.
3. `App::DuckPAN::TemplateDefinitions` is responsible for reading the YAML file and creating one `App::DuckPAN::Template` instance for each template. 
4. `App::DuckPAN::Template` is responsible for creating output files based on variables provided and a list of optional file groups. It also provides a list of output directories that `App::DuckPAN::Restart` watches for changes.

Usage:
duckpan new "Hello World" # creates an instant answer with the 'default' template
duckpan new --template cheatsheet "Hello World" # creates an instant answer with the 'cheatsheet' template

TODO:
1. <s>Tests</s>
2. <s>An option to list all available templates</s>
3. <s>Prompting the user to update the instant answer repository if a `templates.yml` file is not found</s>

(this is part of work related to Issue #268)

zekiel commented:
this is an exciting PR. thanks @srvsh

srvsh commented:
Thanks @zekiel !

I have now updated the PR with test cases and some other very minor modifications. I should be able to get to the pending tasks early next week.

zachthompson commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -44,40 +62,82 @@ sub run {
+
+	# Ask which optional groups to create
+	my @optionals;
+
```
There should be a general prompt for optional files before iterating through individual templates so that they can be skipped entirely, e.g.

> View optional templates?
>      Create $optional file...

srvsh replied:
Done

zachthompson commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -44,40 +62,82 @@ sub run {
+	unless ($template) {
+		$self->app->emit_and_exit(-1,
+			"Template '" . $self->template . "' does not exist for this IA Type: " .
+			$type->{name} . ". Please use one of: " . $available_templates);
```
Let's just say:
```
Available templates:
    cheatsheet
    default
```
Would be great if we could have a short description after each entry.  Perhaps that could be added as another key in the template group?

srvsh replied:
I have updated the _available templates_ messaging. The _description_ key would be nice, but I might leave it for later once we decide on the [new the template file format](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1778), because it is not possible to merge this PR before the template definition files have been merged. Thanks!

zachthompson commented:
Hey @srvsh, I just wanted to touch base on the template changes we discussed in the goodie PR.  Do you want to make those changes as part of this?

srvsh commented:
Hi @zachthompson, yes, I'll send my updated code in this PR within a day. Sorry about the delay; I haven't been able to put in enough time for this yet.

zachthompson commented:
@srvsh, np, just wanted to make sure you weren't waiting on me.

srvsh commented:
@zachthompson I have implemented the new template format that you suggested with some minor modifications. The corresponding YAML file for goodies is [in this pull request](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1778).

Following are some of the terminology I've used:
1. The 'templates.yml' file is called 'the _template definitions file_
2. A _template_ consists of one input and one output file (eg. _pm_, _css_, etc)
3. A _template set_ is a collection of templates representing one instant answer sub-type (eg. for Goodies IA type we have _default_, _cheatsheet_ and _all_ sub-types). The word _sub-type_ is used in messages to the user, but is synonymous to _template set_ in the code. I thought that would be less ambiguous for the user since the word _template_ is used for other things in the documentation.

As per your earlier comment, I have modified the format of the list of the available sub-types. I haven't included the initial prompt to skip viewing optional templates entirely with this set of changes though. (there's a TODO list at the end of this message)

Do let me know what you think!

Here's an example session using `duckpan new` with the changes:
```
srvsh$ duckpan new --list-subtypes
Available sub-types:
           all - Goodie with all backend and frontend files
    cheatsheet - Cheat Sheet Instant Answer
       default - Standard Goodie Instant Answer

srvsh$ duckpan new --subtype cheatsheet
Creating a new Cheat Sheet Instant Answer...
Please enter a name for your Instant Answer:  : readline
Created file: share/goodie/cheat_sheets/json/readline.json
Successfully created Goodie: Readline

srvsh$ duckpan new --subtype all
Creating a new Goodie with all backend and frontend files...
Please enter a name for your Instant Answer:  : weather forecast
Created file: lib/DDG/Goodie/WeatherForecast.pm
Created file: t/WeatherForecast.t
Created file: share/goodie/weather_forecast/weather_forecast.js
Created file: share/goodie/weather_forecast/weather_forecast.css
Successfully created Goodie: WeatherForecast

srvsh$ duckpan new
Creating a new Standard Goodie Instant Answer...
Please enter a name for your Instant Answer:  : bar code
Create Javascript file? [y/n] : y
Create CSS file? [y/n] : n
Created file: lib/DDG/Goodie/BarCode.pm
Created file: t/BarCode.t
Created file: share/goodie/bar_code/bar_code.js
Successfully created Goodie: BarCode
```

Pending tasks:
1. Add a general prompt asking if the user wants to see the optional files
2. Test cases

zachthompson commented:
Hi @srvsh The overall logic looks pretty good.  I have some comments about the new terminology and @moollaza about the flow of `duckpan new`.

Let's stick with `--template` and `--list-templates`.  I think `--subtype` could be confusing to users who don't know the code or config files for duckpan.  They are our primary concern.  Developers tweaking things can be expected to keep `template` in the YAML and `--template` at the command-line distinct.  

 For the output of files created, how about we condense it into:
```
Created files:
    lib/DDG/Goodie/BarCode.pm
    t/BarCode.t
    share/goodie/bar_code/bar_code.js
```
Are the two colons, e.g. `:  : readline`, intentional?

srvsh commented:
@zachthompson thanks for your review and suggestions! I have changed the terminology to use 'template' instead of 'sub-type'. I have also modified the display of the created files. There are a couple of other minor changes renaming some identifiers in my code.

> Are the two colons, e.g. `: : readline`, intentional?

This is not intentional, but the two colons were part of the prompt on my computer before my changes as well. When I run `duckpan new`, this is what I see:
`Please enter a name for your Instant Answer:  : `
I haven't really investigated why it happens though.

I'll send out an update with the test cases today. Thanks!

zachthompson commented on lib/App/DuckPAN.pm:
```diff
@@ -207,6 +207,12 @@ sub get_reply {
 	return $return;
 }
 
+sub ask_yn {
```
Can we not just use [get_reply](https://github.com/srvsh/p5-app-duckpan/blob/generic_templates/lib/App/DuckPAN.pm#L203)?

srvsh replied:
I feel that using `Term::UI::ask_yn` is useful since it does a couple of things automatically to save some code:
1. It adds a '[y/n]' at the end of the prompt with the default option capitalized and
2. Returns a boolean value after parsing the reply

But I understand your concern about defining a sub that is used only once, so I'll go with anything you're comfortable having in the code base.

zachthompson replied:
I agree @srvsh.  Ignore my question!

zachthompson commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -51,33 +103,90 @@ sub run {
+
+	# Ask which optional templates to create
+	my @optional_templates;
+
```
After we ask if they want to configure the optional templates, @moollaza had the idea that we just present a list of optional combinations so they always respond once, e.g.
```
Would you like to configure optional templates? [n]: y

  1> javascript
  2> css
  3> javascript, css
  4> javascript, handlebars
  5> javascript, css, handlebars

Choose configuration [1]: 
```

srvsh replied:
Sounds good, thanks. I'll make this change. I have a clarification though - does the order of the options matter? For example, can `css` be option 1 and `javascript` be 2? And in 5 for example, can the options be `css, handlebars, javascript`?

Should we also expand the idea of showing the menu to listing the templates when the `--template` argument is not passed? Such as:
`
$ duckpan new
Which template would you like to use for your instant answer?

    1> standard (Standard Goodie instant answer with a Perl module and optional frontend files)
    2> cheat sheet (Cheat Sheet with the cheat sheet JSON file)

Choose configuration [1]:
`

zachthompson replied:
I like the idea of making the menus consistent and expanding it to `duckpan new`.

Order is another story.
* For the template(_sets), how about the order in the yaml is the order they are presented with the first being the default?
* For the optional files, how about the order in the array is how they are combined and presented? For example:

`optional: [Javascript, CSS]` produces
```
1> Javascript
2> CSS
3> Javascript, CSS
```
and `optional: [CSS, Javascript]` produces
```
1> CSS
2> Javascript
3> CSS, Javascript
```
This would allow both to be changed easily.  Another idea would be to have a priority value but that seems overly complex.  What do you think?  Any other ideas?

srvsh replied:
I agree with your suggestion - it is intuitive and doesn't need anything changed in the `templates.yml` file other than the ordering. I'll go ahead and implement it this way. Thanks!

zachthompson commented:
Maybe the double colon is due to [Term::UI](https://metacpan.org/pod/Term::UI)? In the default example, if you add a colon to the prompt you get two of them:
``` Perl
my $term = Term::ReadLine->new("brand");
 
my $reply = $term->get_reply(
                prompt => "What is your favourite colour: ",
                choices => [qw|blue red green|],
                default => "blue",
);'

  1> blue
  2> red
  3> green

What is your favourite colour:  [1]:
```

zachthompson commented:
The colon is due to newer Term::UI treating the [empty list](https://github.com/duckduckgo/p5-app-duckpan/blob/master/lib/App/DuckPAN.pm#L205) it's passed differently.  We run version 0.30 internally:
``` Perl
my $term = Term::ReadLine->new("brand");
 
my $reply = $term->get_reply(
                prompt => "What is your favourite colour: ", ()
);'
What is your favourite colour:
```
but on newer versions like 0.46:
``` Perl
my $term = Term::ReadLine->new("brand");
 
my $reply = $term->get_reply(
                prompt => "What is your favourite colour: ",
                ()                              
);'
What is your favourite colour:  : 
```
We could probably check `%params` and only add to the args if true.

srvsh commented:
@zachthompson Ah, you're right. [This commit](https://github.com/jib/term-ui/commit/1e02d71eba7017263520a9eabf6bf3550989445d) in `Term::UI` introduced the new behaviour. From the looks of the it, I'm not sure if anything other than a version check on `Term::UI` can be help. 

Perhaps in `App::DuckPAN::get_reply()` we can check if `$Term::UI::VERSION >= 0.32` and remove the colon and/or spaces at the end. But this gets more messy if the prompt has a `?` at the end instead of a colon, in which case the prompt will look like `What is your username on https://duck.co/ ? : `.

srvsh commented:
@zachthompson I have made changes to show the new way of showing optional templates. Here's a sample session with the changes:

```
# List templates
srvsh$ duckpan new --list-templates
Available templates:
           all - Goodie with all backend and frontend files
    cheatsheet - Cheat Sheet Instant Answer
       default - Standard Goodie Instant Answer

# Default template with optional files
srvsh$ duckpan new vim
Creating a new Standard Goodie Instant Answer...
Would you like to configure optional templates? [y/N]: y

  1> Javascript
  2> CSS
  3> Javascript, CSS

Choose configuration [1]: 3
Created files:
    lib/DDG/Goodie/Vim.pm
    t/Vim.t
    share/goodie/vim/vim.js
    share/goodie/vim/vim.css
Successfully created Goodie: Vim

# Cheat Sheet - no optional files
srvsh$ duckpan new --template cheatsheet elinks
Creating a new Cheat Sheet Instant Answer...
Created files:
    share/goodie/cheat_sheets/json/elinks.json
Successfully created Goodie: Elinks

# On error
srvsh$ duckpan new --template cheatsheet w3m
Creating a new Cheat Sheet Instant Answer...
Created files:
    (none)
[FATAL]  Template output file 'share/goodie/cheat_sheets/json/w3m.json' already exists
```
Does this look alright?

Pending:
1. Test cases (will start on this once the code changes are done)
2. Showing a list of choices for templates instead of choosing a default template when no `--template` option is given. (I'm planning to do this in another PR)

zachthompson commented:
@srvsh I created quite a few new instant answers and mostly looks good.  One edge case:
```
zeroclickinfo-goodies {pr/1778}]$ dpng new --template
Creating a new Standard Goodie Instant Answer...
Would you like to configure optional templates? [y/N]: n
Created files:
    lib/DDG/Goodie/--template.pm
    t/--template.t
Successfully created Goodie: --template
```
What do you think about catching this and displaying a list of templates from which to choose?  We could also just error but really what we want is for them to enter a template name.

Regarding your #2 above about `duckpan new`, let's leave it using the default template.  I think we want a succinct way to do the default with no prompts.

srvsh commented:
Thanks for experimenting with the changes, @zachthompson! I investigated the issue with `--template` and no arguments, and it is in fact a more wider problem affecting all modules. I have submitted #286 with the fix. The issue was that the `bin/duckpan` script changes `Getopt::Long`'s configuration which caused unwanted behaviour in this module.

The behaviour after that fix is that the `duckpan new --template` command chooses the default template and continues execution. It is not intuitive, but sadly that's how `MooX::Options` sets things up for options with default values. I have created an issue with that distribution: https://github.com/celogeek/MooX-Options/issues/49

A workaround should be possible, but I think this case would be rare and since the options would be documented, it may not be too much of an issue. Do you think we can revisit this at a later date?

moollaza commented:
@srvsh @zachthompson regarding the `--template` I believe we just need to add an `option` to the module like we've done in `server.pm`?

https://github.com/duckduckgo/p5-app-duckpan/blob/master/lib/App/DuckPAN/Cmd/Server.pm#L17

zachthompson commented:
@srvsh What I meant about the `--template` is just catching it when it's the sole argument [around here](https://github.com/duckduckgo/p5-app-duckpan/pull/279/files#diff-e62c3a7f663615da614e1882ac30fe4bR114), e.g.
``` Perl
if $self->list_templates || $args[0] eq '--template';
```
Then we at least error with an appropriate message if no template is supplied, rather than creating a `--template.pm`.  In the future we could maybe enter the `select a template dialog` automatically.

srvsh commented:
Sounds good @zachthompson! I've made the change to display a list of templates when `--template` does not have an argument. Here's a sample run with the change:
```
srvsh$ duckpan new --template
[FATAL]  Please specify the template for your Instant Answer.
Available templates:
           all - Goodie with all backend and frontend files
    cheatsheet - Cheat Sheet Instant Answer
       default - Standard Goodie Instant Answer
```
Does that look alright?

zachthompson commented:
:+1:

srvsh commented:
@zachthompson I have made a change to the way the `--template` option with no parameters was handled and added the test cases.

There was an [issue](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1771#issuecomment-161477976) I found earlier which causes output from templates to have empty strings where the `lia_id` variable was used. I had made a temporary fix [here](https://github.com/srvsh/p5-app-duckpan/commit/cd8d8877a75f84d11188b2dc4dcee42da21b26bf) which passes a `lia_id` with the same value as `lia_name`, since that seems to be the same as what the ID should be. Is that a correct assumption?

zachthompson commented:
@srvsh I just reviewed all of the places I could find where we are using lia_id/lia_name/ia_name.  All of these really should be `ia_id` which is the lower-case version of the underscore path.  The only name we use is the separated version.

@moollaza lmk if I missed anything but I think all of these should be ia_id, i.e. this should match the IA page.

zachthompson commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -51,33 +157,100 @@ sub run {
 		lia_name          => $lc_name,
-		ia_path           => $filepath
+		ia_path           => $filepath,
+		lia_id            => $lc_name,
```
1. change to `ia_id`
2. remove `ia_name` and `lia_name` above
3. change templates that use `lia_id`, `lia_name`, and `ia_name` to `ia_id`

srvsh replied:
@zachthompson Thanks, I have made the suggested changes in https://github.com/duckduckgo/p5-app-duckpan/pull/289 , https://github.com/duckduckgo/zeroclickinfo-goodies/issues/1876 , https://github.com/duckduckgo/zeroclickinfo-spice/issues/2331

zachthompson replied:
Excellent.  Have the changes to the templates been made as well?

Just FYI, I'll be out for the next two weeks so @moollaza will be trying to move this forward as time permits.

srvsh replied:
The changes have been made in the current master of the IA repos. The links are in my previous comment. I'm planning to get the changes into the `generic_templates` branch (the branch I've used for generic templating related PRs) once they are reviewed.

> Just FYI, I'll be out for the next two weeks so @moollaza will be trying to move this forward as time permits.
Alright, thanks for letting me know!

moollaza commented:
@srvsh we needed to make some minor updates to DuckPAN this week, can you pull in those changes so this PR is mergeable?I'll be testing this all out and hope to merge it asap.

BTW -- If you haven't responded to any of the automated messages about a free shirt, please do send us an email with your details if you'd like a free shirt! You've done some awesome work and we'd like to thank you :+1:

srvsh commented:
@moollaza Done, I have rebased my changes on top of master so that there are no conflicts.

Thanks for the appreciation :) I did send a mail for a free T-shirt a few today :)

Regarding merging of this PR, I just wanted to let you know that it should be merged after https://github.com/duckduckgo/zeroclickinfo-spice/pull/2295 and https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1778 have been merged to their respective repositories.  The reason is that the code in this PR will error out in older Goodie and Spice repositories due to the missing `templates.yml` file.

srvsh commented:
@moollaza I have incorporated your suggestions from [here](https://github.com/duckduckgo/zeroclickinfo-spice/pull/2295#issuecomment-163712967) into this PR. There are a couple of changes:
1. IA names entered by users are now validated by a basic regex to allow only characters from `[/a-zA-Z0-9\s]`.
2. A new `subdir_support` field has been [added to TemplateSet.pm](https://github.com/duckduckgo/p5-app-duckpan/commit/e5cfa27c52985d2682b70390b6b18a2592e70981#diff-c472809cc457c20f558f0d3c136b85d8R32) to specify which template sets support path separators in their names. The name entered by the user is checked for path separators for template sets that shouldn't have them.
3. The 'success' message has been replaced by `Success!`

Here's a sample `duckpan` session with the changes:
```
srvsh$ duckpan new hello-world
Creating a new Standard Goodie Instant Answer...
[FATAL]  'hello-world' is not a valid name for an Instant Answer. Please run the program again and provide a valid name.

srvsh$ duckpan new duck/duckgo
Creating a new Standard Goodie Instant Answer...
Would you like to configure optional templates? [y/N]: y

  1> Javascript
  2> CSS
  3> Handlebars
  4> Javascript, CSS
  5> Javascript, Handlebars
  6> CSS, Handlebars
  7> Javascript, CSS, Handlebars

Choose configuration [1]: 7
Created files:
    lib/DDG/Goodie/Duck/Duckgo.pm
    t/Duck/Duckgo.t
    share/goodie/duck/duckgo/duck_duckgo.js
    share/goodie/duck/duckgo/duck_duckgo.css
    share/goodie/duck/duckgo/duck_duckgo.handlebars
Success!

srvsh$ duckpan new --template cheatsheet duck/duckhack
Creating a new Cheat Sheet Instant Answer...
[FATAL]  The name for this type of Instant Answer cannot contain path separators. Please run the program again and provide a valid name.

srvsh$ duckpan new --template cheatsheet duck duckhack
Creating a new Cheat Sheet Instant Answer...
Created files:
    share/goodie/cheat_sheets/json/duck_duckhack.json
Success!

```

moollaza commented:
> 1. IA names entered by users are now validated by a basic regex to allow only characters from [/a-zA-Z0-9\s].

We should still allow names to be entered as Perl Packages for non-cheatsheet IA: `duckpan new DDG::Spice::My::Answers` or `duckpan new My::InstantAnswer`

moollaza commented:
Otherwise, this LGTM :+1:

moollaza commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -1,22 +1,112 @@
+##########################
+
+# A 'template' for the user is equivalent to a 'template-set' for the program
+option template => (
```
a shortform, `-t` would be nice --- I'll try and compile of features that can be added later

moollaza commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -1,22 +1,112 @@
+	doc     => 'template used to generate the instant answer skeleton (default: default)',
+);
+
+option list_templates => (
```
likewise I think `--list` would be great` and/or `-L`

moollaza commented on lib/App/DuckPAN/Cmd/New.pm:
```diff
@@ -1,22 +1,112 @@
+	is  => 'ro',
+	doc => 'list the available instant answer templates and exit',
+);
+
```
I'd still like a `--cheatsheet` and/or `-c` option to go straight to a cheat sheet

moollaza commented on t/template/lib/DDG/Default.pm:
```diff
@@ -0,0 +1,2 @@
+package <:$package_name:>;
+# <:$lia_name:>
```
I guess we can remove this comment now?

moollaza commented on t/templates.t:
```diff
@@ -0,0 +1,246 @@
+##################################
+
+my $package_name = 'MyInstantAnswer';
+my $lia_name = 'my_instant_answer';
```
@srvsh I think we need to remove reference to this now? Or change it to `ia_id?`

moollaza commented:
Thanks! This LGTM :+1:

moollaza commented:
@srvsh as I've started using this new code I've found an error. Working with old pull requests that don't have the templates files creates problems. I suspect developers will see issues as well once DuckPAN is released.

The code falls into a loop warning that the templates can't be found -- we need to make it a non-fatal error when duckpan loads as usual, but a fatal error if you use the `duckpan new` function so that developers and i can continue with older branches.

Does that make sense?

srvsh commented:
@moollaza Thanks for your in-depth review! I'll work on your suggestions today.

> The code falls into a loop warning that the templates can't be found -- we need to make it a non-fatal error when duckpan loads as usual, but a fatal error if you use the duckpan new function so that developers and i can continue with older branches.

Could you please give me some steps to reproduce your observation? If I run `duckpan new` in an older goodie/spice repository, I get the fatal error `[FATAL]  Template definitions file not found for Goodie Instant Answers. You may need to pull the latest version of this repository`, after which the program exits. Should I do something else to reproduce the problem you're seeing?

srvsh commented:
@moollaza I've opened PR #294 with a workaround for the problem you were facing. I think that should fix it.
